flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java

```diff
@@ -26,21 +26,22 @@
 
     /**
      * Returns the remaining delay of the processing time specified by {@code processingTimestamp}.
+     * This delay should guarantee that the timer will be fired after all record with timestamp <=
+     * {@code processingTimestamp} have been already processed.
      *
      * @param processingTimestamp the processing time in milliseconds
      * @param currentTimestamp the current processing timestamp; it usually uses {@link
      *     ProcessingTimeService#getCurrentProcessingTime()} to get
      * @return the remaining delay of the processing time
      */
-    public static long getProcessingTimeDelay(long processingTimestamp, long currentTimestamp) {
+    public static long getRecordProcessingTimeDelay(
+            long processingTimestamp, long currentTimestamp) {
 
-        // Two cases of timers here:
-        // (1) future/now timers(processingTimestamp >= currentTimestamp): delay the firing of the
-        //   timer by 1 ms to align the semantics with watermark. A watermark T says we won't see
-        //   elements in the future with a timestamp smaller or equal to T. With processing time, we
-        //   therefore need to delay firing the timer by one ms.
-        // (2) past timers(processingTimestamp < currentTimestamp): do not need to delay the firing
-        //   because currentTimestamp is larger than processingTimestamp pluses the 1ms offset.
+        // future/now timers(processingTimestamp >= currentTimestamp): delay the firing of the
+        // timer by 1 ms to align the semantics with watermark. A watermark T says we won't see
+        // elements in the future with a timestamp smaller or equal to T. Without this 1ms delay,
+        // if we had fired the timer for T at the timestamp T, it would be possible that we would
+        // process another record for timestamp == T in the same millisecond.
```

Comment:
I actually am not sure about these changes by myself so if somebody think that it isn't better than it was before we can just revert these changes.

Reply:
> it would be possible that we would process another record for timestamp == T in the same millisecond.

->

> it would be possible that we would process another record for timestamp == T in the same millisecond, but after the timer for the timsetamp T has already been fired.

Reply:
Plus I would keep the distinction between those two cases (the previous (1) and (2)).